File: flutter/flutter-sdk-bridge/ping_oidc/ios/ping_oidc/Sources/ping_oidc/OidcHostApiImpl.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingBrowser
import PingOidc
import ping_core

/// Implements the generated `PingOidcHostApi`. `configureOidc`/`dispose` (Phase 2),
/// `createWebClient`/`authorize`/`hasUser` (Phase 3), and `token`/`refresh`/`userInfo`/`revoke`/
/// `signOff` (Phase 4) all have real bodies now.
final class OidcHostApiImpl: PingOidcHostApi, @unchecked Sendable {
    func configureOidc(config: OidcConfigMessage, completion: @escaping (Result<String, Error>) -> Void) {
        Task {
            do {
                let handleId = try await OidcClientFactory.create(config)
                completion(.success(handleId))
            } catch {
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.configure, error)))
            }
        }
    }

    func createWebClient(
        clientId: String,
        options: BrowserOptionsMessage?,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        Task {
            do {
                let handleId = try await OidcWebClientFactory.create(clientHandleId: clientId, options: options)
                completion(.success(handleId))
            } catch {
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.createWebClient, error)))
            }
        }
    }

    func authorize(webClientId: String, completion: @escaping (Result<AuthorizeResultMessage, Error>) -> Void) {
        Task {
            guard let handle = await resolveWebClientHandle(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.authorize, OidcHostApiError.stateError(
                        "OIDC web client instance not found for id=\(webClientId)"
                    )
                )))
                return
            }
            do {
                let result = try await handle.webClient.authorize()
                switch result {
                case .success:
                    completion(.success(AuthorizeResultMessage(type: .success)))
                case .failure(let error):
                    completion(mapAuthorizeFailure(error))
                }
            } catch {
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.authorize, error)))
            }
        }
    }

    /// A cancelled `ASWebAuthenticationSession` must resolve `AuthorizeResultMessage.type` as
    /// `.cancel`, not surface as a thrown/mapped error.
    ///
    /// TODO(SDKS-5295): ping-ios-sdk's `OidcWebClient.authorize()` collapses every
    /// `FailureNode.cause` into `OidcError.unknown(message:)`, discarding
    /// `BrowserError.externalUserAgentCancelled`'s type entirely (confirmed via source trace —
    /// `Web.swift`'s bare `catch` rewraps it as a string before it ever reaches here). This
    /// substring match on the resulting message is the only signal available today. It is
    /// fragile — it silently stops working if the SDK ever rewords the message — and should be
    /// removed once the native SDK preserves a distinguishable cancellation signal, tracked in
    /// SDKS-5295 to align it with Android's typed `BrowserCanceledException`.
    private func mapAuthorizeFailure(_ error: OidcError) -> Result<AuthorizeResultMessage, Error> {
        if plainErrorMessage(for: error).contains("cancelled by the user") {
            return .success(AuthorizeResultMessage(type: .cancel))
        }
        return .failure(OidcErrorMapper.from(OidcErrorCodes.authorize, error))
    }

    func hasUser(webClientId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        Task {
            guard let handle = await resolveWebClientHandle(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.hasUser, OidcHostApiError.stateError(
                        "OIDC web client instance not found for id=\(webClientId)"
                    )
                )))
                return
            }
            let user = await handle.webClient.user()
            completion(.success(user != nil))
        }
    }

    func token(webClientId: String, completion: @escaping (Result<TokenMessage, Error>) -> Void) {
        Task {
            guard let user = await resolveUser(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.token, OidcHostApiError.stateError(
                        "No OIDC user session for id=\(webClientId)"
                    )
                )))
                return
            }
            switch await user.token() {
            case .success(let token):
                completion(.success(token.toMessage()))
            case .failure(let error):
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.token, error)))
            }
        }
    }

    // Go through `User.refresh()` on both platforms — iOS has no `OidcClient.refresh()`
    // equivalent, so this is also the only option, not just the preferred one.
    func refresh(webClientId: String, completion: @escaping (Result<TokenMessage, Error>) -> Void) {
        Task {
            guard let user = await resolveUser(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.refresh, OidcHostApiError.stateError(
                        "No OIDC user session for id=\(webClientId)"
                    )
                )))
                return
            }
            switch await user.refresh() {
            case .success(let token):
                completion(.success(token.toMessage()))
            case .failure(let error):
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.refresh, error)))
            }
        }
    }

    func userInfo(
        webClientId: String,
        cache: Bool,
        completion: @escaping (Result<[String?: Any?], Error>) -> Void
    ) {
        Task {
            guard let user = await resolveUser(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.userInfo, OidcHostApiError.stateError(
                        "No OIDC user session for id=\(webClientId)"
                    )
                )))
                return
            }
            // `cache` is always passed through explicitly — Android's own default is `false`,
            // iOS's is `true`; never rely on either platform default.
            switch await user.userinfo(cache: cache) {
            case .success(let info):
                let bridged: [String?: Any?] = info.reduce(into: [:]) { result, entry in
                    result[entry.key] = entry.value
                }
                completion(.success(bridged))
            case .failure(let error):
                completion(.failure(OidcErrorMapper.from(OidcErrorCodes.userInfo, error)))
            }
        }
    }

    func revoke(webClientId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            guard let user = await resolveUser(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.revoke, OidcHostApiError.stateError(
                        "No OIDC user session for id=\(webClientId)"
                    )
                )))
                return
            }
            await user.revoke()
            completion(.success(()))
        }
    }

    // Known SDK gap (see IMPLEMENTATION_PLAN_OIDC.md Phase 4): the only reachable sign-off path
    // from a web-client id is `User.logout()`, which returns `Void` and discards the underlying
    // `signOff()` result — the boolean-returning `OidcClient.endSession()` is not reachable here.
    // `true` below means "logout completed without a bridge-level exception," not "a session
    // existed and was terminated."
    func signOff(webClientId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        Task {
            guard let user = await resolveUser(webClientId) else {
                completion(.failure(OidcErrorMapper.from(
                    OidcErrorCodes.signOff, OidcHostApiError.stateError(
                        "No OIDC user session for id=\(webClientId)"
                    )
                )))
                return
            }
            await user.logout()
            completion(.success(true))
        }
    }

    func dispose(handleId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            // A dispose call may target either a client handle or a web-client handle; removal
            // is a no-op on an unknown id in both registries (SimpleRegistry.remove), so trying
            // both unconditionally is safe and avoids tracking which registry a given id came
            // from.
            await CoreRuntime.oidcClientRegistry.remove(handleId)
            await CoreRuntime.oidcWebClientRegistry.remove(handleId)
            completion(.success(()))
        }
    }

    private func resolveWebClientHandle(_ webClientId: String) async -> OidcWebClientHandle? {
        await CoreRuntime.oidcWebClientRegistry.resolve(webClientId) as? OidcWebClientHandle
    }

    /// Resolves `webClientId` to a live `User`, or `nil` if no session exists yet.
    private func resolveUser(_ webClientId: String) async -> User? {
        guard let handle = await resolveWebClientHandle(webClientId) else { return nil }
        return await handle.webClient.user()
    }
}

/// Maps a native `Token` onto the bridge's `TokenMessage` — fields match 1:1; `expiresAt`
/// (native's own expiry stamp) has no `TokenMessage` counterpart and is dropped.
private extension Token {
    func toMessage() -> TokenMessage {
        TokenMessage(
            accessToken: accessToken,
            tokenType: tokenType,
            scope: scope,
            expiresIn: expiresIn,
            refreshToken: refreshToken,
            idToken: idToken
        )
    }
}

File: flutter/flutter-sdk-bridge/ping_oidc/android/src/main/kotlin/com/pingidentity/flutter/oidc/OidcHostApiImpl.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.flutter.oidc

import com.pingidentity.browser.BrowserCanceledException
import com.pingidentity.flutter.core.CoreRuntime
import com.pingidentity.flutter.core.mapper.JsonBridgeMapper
import com.pingidentity.flutter.oidc.error.OidcErrorCodes
import com.pingidentity.flutter.oidc.error.OidcErrorMapper
import com.pingidentity.flutter.oidc.error.classifyError
import com.pingidentity.oidc.OidcError
import com.pingidentity.oidc.Token
import com.pingidentity.oidc.User
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import com.pingidentity.utils.Result as OidcResult

/**
 * Implements the generated [PingOidcHostApi]. `configureOidc`/`dispose` (Phase 2),
 * `createWebClient`/`authorize`/`hasUser` (Phase 3), and `token`/`refresh`/`userInfo`/`revoke`/
 * `signOff` (Phase 4) all have real bodies now.
 */
class OidcHostApiImpl : PingOidcHostApi {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    override fun configureOidc(config: OidcConfigMessage, callback: (Result<String>) -> Unit) {
        callback(
            runCatching { OidcClientFactory.create(config) }
                .classifyError(OidcErrorCodes.CONFIGURE)
        )
    }

    override fun createWebClient(
        clientId: String,
        options: BrowserOptionsMessage?,
        callback: (Result<String>) -> Unit
    ) {
        callback(
            runCatching { OidcWebClientFactory.create(clientId, options) }
                .classifyError(OidcErrorCodes.CREATE_WEB_CLIENT)
        )
    }

    override fun authorize(webClientId: String, callback: (Result<AuthorizeResultMessage>) -> Unit) {
        // Custom Tab launch needs the main thread and differs from
        // this class's other calls, which run on Dispatchers.IO.
        scope.launch(Dispatchers.Main) {
            val outcome =
                runCatching { resolveWebClientHandle(webClientId).webClient.authorize() }
                    .fold(
                        onSuccess = { authorizeResult ->
                            authorizeResult.fold(
                                onSuccess = {
                                    Result.success(AuthorizeResultMessage(AuthorizeResultType.SUCCESS))
                                },
                                onFailure = { cause -> mapAuthorizeFailure(cause) },
                            )
                        },
                        onFailure = { error -> mapAuthorizeFailure(error) },
                    )
            callback(outcome)
        }
    }

    /**
     * A cancelled Custom Tab must resolve [AuthorizeResultMessage.type] as `cancel`, not surface
     * as a thrown/classified error — this check runs before [OidcErrorMapper] so cancellation
     * never reaches it.
     */
    private fun mapAuthorizeFailure(cause: Throwable): Result<AuthorizeResultMessage> =
        if (cause is BrowserCanceledException) {
            Result.success(AuthorizeResultMessage(AuthorizeResultType.CANCEL))
        } else {
            Result.failure(OidcErrorMapper.fromThrowable(OidcErrorCodes.AUTHORIZE, cause))
        }

    override fun hasUser(webClientId: String, callback: (Result<Boolean>) -> Unit) {
        scope.launch {
            val result =
                runCatching { resolveWebClientHandle(webClientId).webClient.user() != null }
            callback(result.classifyError(OidcErrorCodes.HAS_USER))
        }
    }

    override fun token(webClientId: String, callback: (Result<TokenMessage>) -> Unit) {
        scope.launch {
            callback(fetchToken(webClientId, OidcErrorCodes.TOKEN) { it.token() })
        }
    }

    override fun refresh(webClientId: String, callback: (Result<TokenMessage>) -> Unit) {
        // Go through `User.refresh()` on both platforms, not `OidcClient.refresh()` — iOS has no
        // equivalent, so this keeps one code shape (see IMPLEMENTATION_PLAN_OIDC.md § Platform
        // asymmetries).
        scope.launch {
            callback(fetchToken(webClientId, OidcErrorCodes.REFRESH) { it.refresh() })
        }
    }

    /** Shared body for [token]/[refresh]: resolve the user, call [tokenCall], map the result. */
    private suspend fun fetchToken(
        webClientId: String,
        code: String,
        tokenCall: suspend (User) -> OidcResult<Token, OidcError>,
    ): Result<TokenMessage> =
        runCatching { resolveUser(webClientId) }
            .fold(
                onSuccess = { user ->
                    when (val result = tokenCall(user)) {
                        is OidcResult.Success -> Result.success(result.value.toMessage())
                        is OidcResult.Failure ->
                            Result.failure(OidcErrorMapper.fromOidcError(code, result.value))
                    }
                },
                onFailure = { error -> Result.failure(OidcErrorMapper.fromThrowable(code, error)) },
            )

    override fun userInfo(
        webClientId: String,
        cache: Boolean,
        callback: (Result<Map<String?, Any?>>) -> Unit
    ) {
        scope.launch {
            val outcome =
                runCatching { resolveUser(webClientId) }
                    .fold(
                        onSuccess = { user ->
                            // `cache` is always passed through explicitly — Android's own default
                            // is `false`, iOS's is `true`; never rely on either platform default.
                            when (val result = user.userinfo(cache)) {
                                is OidcResult.Success -> {
                                    val flattened =
                                        JsonBridgeMapper.encodeJsonElement(result.value)
                                            as? Map<String?, Any?> ?: emptyMap()
                                    Result.success(flattened)
                                }
                                is OidcResult.Failure ->
                                    Result.failure(
                                        OidcErrorMapper.fromOidcError(OidcErrorCodes.USERINFO, result.value)
                                    )
                            }
                        },
                        onFailure = { error ->
                            Result.failure(OidcErrorMapper.fromThrowable(OidcErrorCodes.USERINFO, error))
                        },
                    )
            callback(outcome)
        }
    }

    override fun revoke(webClientId: String, callback: (Result<Unit>) -> Unit) {
        scope.launch {
            val outcome = runCatching { resolveUser(webClientId).revoke() }
                .classifyError(OidcErrorCodes.REVOKE)
            callback(outcome)
        }
    }

    override fun signOff(webClientId: String, callback: (Result<Boolean>) -> Unit) {
        // Known SDK gap (see IMPLEMENTATION_PLAN_OIDC.md Phase 4): the only reachable sign-off
        // path from a web-client id is `User.logout()`, which returns `Unit` and discards the
        // underlying `Workflow.signOff()` result — the boolean-returning `OidcClient.endSession()`
        // is not reachable here. `true` below means "logout completed without a bridge-level
        // exception," not "a session existed and was terminated."
        scope.launch {
            val outcome =
                runCatching { resolveUser(webClientId).logout() }
                    .fold(
                        onSuccess = { Result.success(true) },
                        onFailure = { error ->
                            Result.failure(OidcErrorMapper.fromThrowable(OidcErrorCodes.SIGN_OFF, error))
                        },
                    )
            callback(outcome)
        }
    }

    override fun dispose(handleId: String, callback: (Result<Unit>) -> Unit) {
        scope.launch {
            // A dispose call may target either a client handle or a web-client handle; removal is
            // a no-op on an unknown id in both registries (SimpleRegistry.remove), so trying both
            // unconditionally is safe and avoids tracking which registry a given id came from.
            val result =
                runCatching {
                    CoreRuntime.oidcClientRegistry.remove(handleId)
                    CoreRuntime.oidcWebClientRegistry.remove(handleId)
                }
            callback(result.classifyError(OidcErrorCodes.DISPOSE))
        }
    }

    private fun resolveWebClientHandle(webClientId: String): OidcWebClientHandle =
        (CoreRuntime.oidcWebClientRegistry.resolve(webClientId) as? OidcWebClientHandle)
            ?: throw IllegalStateException(
                "OIDC web client instance not found for id=$webClientId"
            )

    /** Resolves [webClientId] to a live [User], or throws if no session exists yet. */
    private suspend fun resolveUser(webClientId: String): User =
        resolveWebClientHandle(webClientId).webClient.user()
            ?: throw IllegalStateException("No OIDC user session for id=$webClientId")

    /** Cancels all in-flight coroutines. Called from [PingOidcPlugin.onDetachedFromEngine]. */
    fun shutdown() {
        scope.cancel()
    }
}

/** Maps a native [Token] onto the bridge's [TokenMessage] — fields match 1:1; `expireAt` (native's
 * internal expiry stamp) has no [TokenMessage] counterpart and is dropped. */
private fun Token.toMessage(): TokenMessage =
    TokenMessage(
        accessToken = accessToken,
        tokenType = tokenType,
        scope = scope,
        expiresIn = expiresIn,
        refreshToken = refreshToken,
        idToken = idToken,
    )
